Replace `ExternalStream` with new `ByteStream` type

IanManske opened the pull request:
# Description
This PR introduces a `ByteStream` type which is a `Read`-able stream of bytes. Internally, it has an enum over three different byte stream sources:
```rust
pub enum ByteStreamSource {
    Read(Box<dyn Read + Send + 'static>),
    File(File),
    Child(ChildProcess),
}
```

This is in comparison to the current `RawStream` type, which is an `Iterator<Item = Vec<u8>>` and has to allocate for each read chunk.

Currently, `PipelineData::ExternalStream` serves a weird dual role where it is either external command output or a wrapper around `RawStream`. `ByteStream` makes this distinction more clear (via `ByteStreamSource`) and replaces `PipelineData::ExternalStream` in this PR:
```rust
pub enum PipelineData {
    Empty,
    Value(Value, Option<PipelineMetadata>),
    ListStream(ListStream, Option<PipelineMetadata>),
    ByteStream(ByteStream, Option<PipelineMetadata>),
}
```

The PR is relatively large, but a decent amount of it is just repetitive changes.

This PR fixes #7017, fixes #10763, and fixes #12369.

This PR also improves performance when piping external commands. Nushell should, in most cases, have competitive pipeline throughput compared to, e.g., bash.
| Command                                            | Before (MB/s) | After (MB/s) | Bash (MB/s) |
| -------------------------------------------------- | -------------:| ------------:| -----------:|
| `throughput \| rg 'x'`                             | 3059          | 3744         | 3739        |
| `throughput \| nu --testbin relay o> /dev/null`    | 3508          | 8087         | 8136        |

# User-Facing Changes
- This is a breaking change for the plugin communication protocol, because the `ExternalStreamInfo` was replaced with `ByteStreamInfo`. Plugins now only have to deal with a single input stream, as opposed to the previous three streams: stdout, stderr, and exit code.
- The output of `describe` has been changed for external/byte streams.
- Temporary breaking change: `bytes starts-with` no longer works with byte streams. This is to keep the PR smaller, and `bytes ends-with` already does not work on byte streams.
- If a process core dumped, then instead of having a `Value::Error` in the `exit_code` column of the output returned from `complete`, it now is a `Value::Int` with the negation of the signal number.

# After Submitting
- Update docs and book as necessary
- Release notes (e.g., plugin protocol changes)
- Adapt/convert commands to work with byte streams (high priority is `str length`, `bytes starts-with`, and maybe `bytes ends-with`).
- Refactor the `tee` code, Devyn has already done some work on this.

sholderbach commented:
That is an impressive chunk of work and a nice improvement!
Resolving those incorrect buffering bugs is long overdue.

Before we run ahead just looking at the streaming side we need to talk about the error changes, to ensure we don't break our execution semantics in unintended ways. 

- What is the motivation for turning the exit codes into `ShellError` variants? (The existing separate `ListStream` certainly feels dirty, but the semantics change in major ways if it is an explicit error passing around but we have to look at all the relevant places, and I haven't gotten to that yet)
- Can they be isolated into a separate PR in an easy way or will this become a furball here?
- As externals are built in the world where bash is run without `set -e` exit codes are non-zero in ways that may not be errors in their logic. We treat them as such in complex executions but have to tolerate them in individual repl executions for example. See 
@sophiajt 's blog [post on examples](https://www.sophiajt.com/exit-codes/)
- Destructuring `ShellErrors` in `catch` is still not smoothly possible, how would this affect the UX?


>  After Submitting
> 
>     * Figure out end line trimming

What is up here? Is there a breakage we need to resolve before merging?

WindSoilder commented:
I have the same concern about turning the exit codes into `ShellError` variants.
I remember @elferherrera and @sophiajt vote against it before: https://github.com/nushell/nushell/pull/6079#issuecomment-1189168717
https://github.com/nushell/nushell/pull/6079#issuecomment-1189518033

devyn commented on crates/nu-protocol/src/io/lines.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+pub struct Values<R: Read> {
+    lines: ByteLines<R>,
```
This is something I've already changed on my branch, but just so you know, I don't think `Values` should really be lines. I think it's currently just whatever the buffers coming in are, and it's additionally split up by lines to kind of do a pseudo-line-buffering, but that's a source of bugs.

Changing it to always be the full line is okay for text data, but not ideal for binary data because the chunks are all going to be random sizes / potentially the whole file because you never really know when you'll get a `\n` character.

If you want to check out my branch instead, I made it basically use `BufRead::fill_buf` to get whatever data is pending and try to create valid UTF-8 buffers when in string mode. It's quite obviously different from the output of `lines`, so that should help to force anyone who actually wants lines to use that.

IanManske replied:
Hmm, yeah the `fill_buf` approach does sound better. Ideally, `values` should not be needed / should not exist on `ByteStream`, and commands would just use the reader directly (or maybe `lines`). Unfortunately, we have a bunch of commands that don't do that and/or use one of the `PipelineData` methods (`map`, `filter`, `flat_map`, `into_iter_strict`, the `IntoIterator` trait, etc.).

devyn commented on crates/nu-protocol/src/io/lines.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+impl<R: Read> Iterator for Lines<R> {
+    type Item = Result<Vec<u8>, ShellError>;
```
If you're dealing with lines, you're basically always dealing with UTF-8 text anyway (other encodings would potentially have different newline sequences), so in my case I rolled the parsing logic into this iterator and just had it spit out strings. I left `ByteLines` though in case someone really does want that.

devyn commented on crates/nu-command/src/filters/lines.rs:
```diff
@@ -76,27 +110,15 @@ impl Command for Lines {
-                        src_span: val.span(),
-                    }),
+            PipelineData::ByteStream(stream, ..) => {
+                if let Some(lines) = stream.values() {
```
Just in case you do decide to change `Values`, keep in mind this needs to change too

devyn commented on crates/nu-cmd-lang/src/core_commands/do_.rs:
```diff
@@ -37,11 +36,6 @@ impl Command for Do {
                 "ignore external program errors as the closure runs",
                 Some('p'),
             )
-            .switch(
```
In case we do decide to keep the exit codes as `ShellError`, I think you need to change this to a deprecated flag rather than removing it. I have `do -c` in my config file and it fails to parse

devyn commented on crates/nu-plugin-protocol/src/lib.rs:
```diff
@@ -82,32 +82,20 @@ pub enum PipelineDataHeader {
 }
 
 impl PipelineDataHeader {
     /// Return a list of stream IDs embedded in the header
```
```suggestion
    /// Return the stream ID, if any, embedded in the header
```

devyn commented on crates/nu-protocol/src/io/read_iterator.rs:
Because these are quite commonly used with `ByteStream`, I'm thinking of adding something like `ByteStream::from_iter()` and `ByteStream::from_try_iter()` (?) so that you don't have to construct these manually. Thoughts?

IanManske replied:
Oh yeah, that's a good idea!

devyn commented on crates/nu-cmd-lang/src/core_commands/describe.rs:
```diff
@@ -162,59 +162,30 @@ fn run(
                 )
             } else {
-                Value::string("raw input", head)
+                stream.drain()?;
```
probably shouldn't drain if `no_collect`

IanManske commented:
> Can they be isolated into a separate PR in an easy way or will this become a furball here?

Actually, it looks like it might not be too bad to separate it.

>  > *Figure out end line trimming
>
> What is up here? Is there a breakage we need to resolve before merging?

Whoops, that's a TODO, not sure why I put it under after submitting.

IanManske commented:
Ok, I've split off #12782 and #12799 to make this PR slightly smaller. I've also removed the exit code changes. Otherwise, I just have to add doc comments and maybe some tests.

devyn commented:
I just gave it a test and while the error is gone and `complete` returns exit codes, I don't seem to be seeing `LAST_EXIT_CODE` set even for very simple commands on the prompt.

devyn commented on crates/nu-plugin-core/src/interface/mod.rs:
```diff
@@ -185,31 +180,10 @@ pub trait InterfaceManager {
-                    trim_end_newline: info.trim_end_newline,
-                }
+                let reader = handle.read_stream(info.id, self.get_interface())?;
+                ByteStream::from_result_iter(reader, info.span, None).into()
```
In this case, `from_result_iter()` does need to take `ctrlc`, because the underlying stream is not checking it.

IanManske commented:
Ok, I think this should be mostly ready for review.

Also, I stole some of your tests @devyn :P

devyn commented:
I'll have a look again, then. Thanks for all of your hard work!

WindSoilder reviewed (commented):
Thanks for the improvement!  Just leaves some thought about stderr consumer thread.

sholderbach commented:
Is #12369 still part of this after removing the main error breaking changes?

IanManske commented:
Yes, this PR still fixes #12369. If there is no exit code (i.e., the process was terminated via signal), then `complete` will now report the negation of the signal number in `exit_code`.

sholderbach commented on crates/nu-cmd-extra/tests/commands/bytes/starts_with.rs:
```diff
@@ -19,102 +19,102 @@ fn basic_string_fails() {
-    );
-    assert_eq!(actual.out, "false");
-}
+// #[test]
+// fn short_stream_binary() {
+//     let actual = nu!(r#"
+//             nu --testbin repeater (0x[01]) 5 | bytes starts-with 0x[010101]
```
Pin: commented out tests

stormasm commented:
@devyn so we agreed in the core team meeting today that once you give the green light and have reviewed the PR we will go ahead and land it...

Thank you kindly for doing this ! 😄

devyn commented on crates/nu-command/src/bytes/starts_with.rs:
```diff
@@ -60,63 +60,13 @@ impl Command for BytesStartsWith {
-                for item in stream {
-                    let byte_slice = match &item {
-                        // String and binary data are valid byte patterns
-                        Ok(Value::String { val, .. }) => val.as_bytes(),
```
Removing this will be a breaking change, as this worked before:

```nushell
^echo foo | bytes starts-with ("foo" | into binary)
```

In general, I think untyped byte streams should be able to be treated as binary

IanManske replied:
That was me being ~~lazy~~ making the PR smaller :smile:. After this, I intend to open another PR to restore that command and also make `bytes ends-with` support byte streams properly.

devyn commented on crates/nu-command/src/filesystem/save.rs:
```diff
@@ -121,40 +119,102 @@ impl Command for Save {
+                // https://github.com/nushell/nushell/pull/9377 contains the reason
+                // for not using BufWriter<File>
+                //
+                // TODO: flag to specify buffered (BufWriter), line buffered (LineWriter), or no buffering (File).
```
I'm not sure this will ever be necessary, because the input stream is generally buffered in some way, even if by the command that produced it. If someone knows they want to use bigger buffers, it might be better to provide a command to adapt the stream with `read_exact()` or `read_until()` calls to ensure the desired chunking behaviour. And of course it should already be possible to do `| lines | save` if you definitely want line buffering.

I like this behaviour though, it means that the output of an external command gets written to file with the flushing behaviour it was intended to have, which is particularly good if saving to something like a FIFO where the interactivity actually matters.

IanManske replied:
Oh yeah, that's a good observation! There might be an argument to add the flags for performance reasons (e.g., to avoid the allocations), but it's probably not worth it / it won't make a noticeable difference.

devyn reviewed (commented):
Looks good. I found one potential issue with `bytes starts-with`. A lot of the files changed are just `into_value()` related, with that being a `Result` now, so this isn't quite as big as it looks.

Didn't have any issues running it yet, and exit code is definitely working now. I'm glad you changed the `each` behaviour to `Chunks` and got rid of implicit line buffering.

Great stuff!

IanManske commented:
Thanks for taking the time to review this @devyn!

stormasm commented:
@IanManske I am planning on landing this PR in the morning --- its 10pm in Oregon now...

Unless someone beats me to it...

I checked out your branch and ran all of the tests and it looks good to me...

I will await you giving me the green light to let me know if its now ready to go based on Devyn's feedback it looks good to me...

Thanks for all of your hard work on this PR...

IanManske commented:
Sounds good, feel free to land it whenever @stormasm. Thanks!

stormasm commented:
@IanManske cool !  thanks for giving the green light 👍 
I will monitor and make sure I don't see any issues now that its landed....

rgwood commented:
Wow, it’s great to see thoughtful architectural changes like this. Big thanks to Ian and everyone who reviewed this.

YizhePKU commented:
After wrapping a child process into a `ByteStream`, how do I read from its stderr? The public APIs all seem to be reading from stdout.

I'm currently doing this, but it feels awkward:

```rust
if let PipelineData::ByteStream(byte_stream, ..) = pipeline_data {
    if let ByteStreamSource::Child(child) = byte_stream.source() {
        if let Some(stderr) = child.stderr {
            stderr.read(&mut buf);
        }
    }
}
```

YizhePKU commented:
```rust
impl ByteStream {
    pub fn write_to(self, dest: &mut impl Write) -> Result<Option<ExitStatus>, ShellError> {
       // ...
    }
}
```

Generic reader/writer functions should take `R: Read` and `W: Write` by value ([Rust API Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-rw-value)).

IanManske commented:
> After wrapping a child process into a ByteStream, how do I read from its stderr? The public APIs all seem to be reading from stdout.

This is intentionally verbose. In fact, I should have maybe made it an unsafe API. Reading from `stderr` if `stdout` is present and vice versa could create a deadlock (unless, e.g., a separate thread is used to read `stdout` at the same time). In the vast majority of cases, the child process should not have a `stderr`. Commands that set the `stderr` destination to `OutDest::Capture` make sure to read `stdout` and `stderr` on separate threads. To "read" both `stdout` and `stderr` you can use `ChildProcess::wait_with_output`. Hope this helps.

> Generic reader/writer functions should take `R: Read` and `W: Write` by value ([Rust API Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-rw-value)).

Oh, good to know. I guess `io::copy` was an unfortunate oversight?

YizhePKU commented:
> This is intentionally verbose. In fact, I should have maybe made it an unsafe API.

Rust's safety guarantees doesn't cover deadlocks, so it's probably fine.

> I guess io::copy was an unfortunate oversight?

Unfortunate indeed. The Rust team [tried to change it](https://github.com/rust-lang/rust/issues/105643), but couldn't due to API breakage.

Kixunil commented:
> Generic reader/writer functions should take `R: Read` and `W: Write` by value ([Rust API Guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-rw-value)).

FYI this tends to lead to bigger binary sizes because people are using both variants. If you pass by mut ref they will make owned types into references.
